bundles/org.palladiosimulator.simexp.core/src/org/palladiosimulator/simexp/core/valuefunction/MonteCarloPrediction.java

```diff
@@ -0,0 +1,154 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+import org.palladiosimulator.simexp.core.entity.DefaultSimulatedExperience;
+import org.palladiosimulator.simexp.core.entity.SimulatedExperience;
+import org.palladiosimulator.simexp.core.evaluation.SampleModelIterator;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+
+public class MonteCarloPrediction implements ValueFunctionEstimator {
+
+	private abstract class MonteCaroEstimator implements BiConsumer<Set<String>, List<SimulatedExperience>> {
+			
+		@Override
+		public void accept(Set<String> t, List<SimulatedExperience> u) {
+			estimate(t,u);
+		}
+		
+		public abstract void estimate(Set<String> t, List<SimulatedExperience> u);
+		
+	}
+	
+	private static class AccumulatedReward {
+		
+		private double total = 0;
+		private int accummulations = 0;
+		
+		public static AccumulatedReward nonAccumulations() {
+			AccumulatedReward accReward = new AccumulatedReward();
+			accReward.accummulations = 1;
+			return accReward;
+		}
+		
+		public void append(double reward) {
+			total += reward;
+			accummulations++;
+		}
+		
+		public void mergeWith(AccumulatedReward accReward) {
+			total += accReward.total;
+			accummulations += accReward.accummulations;
+		}
+		
+		public double calculateAverage() {
+			return total / accummulations;
+		}
+		
+	}
+	
+	private static class AccumulatedRewardManager {
+		
+		private final Map<String, AccumulatedReward> accRewards = Maps.newHashMap();
+		
+		public void append(String state, AccumulatedReward accReward) {
+			AccumulatedReward accRewardToUpdate = accRewards.get(state);
+			if (accRewardToUpdate == null) {
+				accRewards.put(state, accReward);
+			} else {
+				accRewardToUpdate.mergeWith(accReward);
+			}
+		}
+		
+		public AccumulatedReward getAccumulatedRewardFor(String state) {
+			return Optional.ofNullable(accRewards.get(state)).orElse(AccumulatedReward.nonAccumulations());
+		}
+		
+	}
+	
+	private final ValueFunction valueFunction;
+	private final AccumulatedRewardManager accRewardManager;
+	
+	private MonteCaroEstimator predictionEstimator;
+	
+	private MonteCarloPrediction() {
+		this.valueFunction = new ValueFunction();
+		this.accRewardManager = new AccumulatedRewardManager();
+	}
+	
+	public static MonteCarloPrediction firstVisitEstimation() {
+		MonteCarloPrediction estimation = new MonteCarloPrediction();
+		estimation.predictionEstimator = estimation.firstVisitEstimator();
+		return estimation;
+	}
+	
+	@Override
+	public ValueFunction estimate(SampleModelIterator iterator) {	
+		while(iterator.hasNext()) {
+			List<SimulatedExperience> traj = iterator.next();
+			
+			Set<String> distinctTrajStates = traj.stream()
+					.map(each -> DefaultSimulatedExperience.getCurrentStateFrom(each))
+					.distinct()
+					.collect(Collectors.toSet());
+			
+			predictionEstimator.estimate(distinctTrajStates, traj);			

```

Comment:
wo ist die eigentliche 'estimate' funktionalität hinterlegt? Ich finde den Code an dieser Stelle sehr schwer verständlich über die eingebaute Indirektion von abstract class MonteCaroEstimator implements BiConsumer... ist diese Komplexität hier zwingend notwendig? Meines Erachtens wäre ein entsprechender simpler methodenaufruf völlig ausreichend.

Reply:
Es gibt zwei Varianten von Monte Carlo Prediction, die first-visit und every-visit Methode. Das drückt der Code hier an der Stelle aus (nur das die every-visit Methode nicht implementiert ist). Der entsprechende Code ist in der firstVisitEstimator-Methode zu finden.

Reply:
Auch hier s. Argument von oben. Wenn man als interface MonteCarloPrediction hat mit einer estimate methode, dann kann man diese für beide Alternativen (First/Every) in der jeweiligen Klasse implementieren. Das estimate an sich ändert sich dadurch nicht und man müsste sich dafür auch keine 2 unterschiedlichen Methodennamen einfürhen.  Das grundsätzliuche Konzept ist 'estimate' und nicht das Verfahren wie man es erreicht (first/every), dass wäre dann nur die beiden Ausprägungen der jeweiligen Klasse.